src/SimplePromiseQueue.ts

```diff
@@ -1,25 +1,26 @@
 // tslint:disable: variable-name
 export default class SimplePromiseQueue {
   private readonly _queue: Array<() => Promise<void>> = []
-  private _flushing = false
+  private _runner: Promise<void> | null = null
 
   public enqueue(task: () => Promise<void>) {
     this._queue.push(task)
-    if (!this._flushing) { return this.flushQueue() }
-    return Promise.resolve()
+    this.join()
   }
 
-  public flushQueue() {
-    this._flushing = true
-
-    const chain = (): Promise<void> | void => {
-      const nextTask = this._queue.shift()
-      if (nextTask) {
-        return nextTask().then(chain)
-      } else {
-        this._flushing = false
+  public async join(): Promise<void> {
+    const runner = async () => {
+      while (this._queue.length) {
+        const nextTask = this._queue.shift()!
+        await nextTask()
       }
     }
-    return Promise.resolve(chain())
+
+    if (this._runner) {
+      return await this._runner
+    }
+    this._runner = runner()
+    await this._runner
+    this._runner = null
   }
 }
```

Comment:
Another way that creates an explicit promise.

```javascript
  private _joinPromise: Promise<void> | null = null
  ...
  public async join(): Promise<void> {
    if (!this._joinPromise) {
      this._joinPromise = new Promise((resolve) => {
        while (this._queue.length) {
          const nextTask = this._queue.shift()!
          await nextTask()
        }
        this._joinPromise = null;
        resolve()
      })
    }
    await this._joinPromise
  }
```

Reply:
Clearing the promise in the function is a good idea to avoid race conditions. I changed that but left the function the same. I personally think the `async` function is clearer than creating a Promise and from my understanding they are the same.